concurrency/MpmcPublish.h

```diff
@@ -14,82 +16,77 @@ class MpmcPublish
     {
         Entry() : refCount(0) {}
 
-        std::atomic_uint32_t refCount;
+        std::atomic_uint32_t refCount; // 0 means free
         T value;
     };
 
-    static constexpr uint32_t indexBits = 16;
-    static constexpr uint32_t TOO_MANY_THREADS = 1 << (indexBits - 1);
-    static constexpr uint32_t indexMask = (1 << indexBits) - 1;
-
-    uint32_t indexOf(uint32_t cursor) const { return cursor & indexMask; }
-    uint32_t versionOf(uint32_t cursor) const { return cursor >> indexBits; }
-    uint32_t toCursor(uint32_t version, uint32_t index) const { return (version << indexBits) | 0x80000000 | index; }
-
 public:
     typedef T ValueType;
-    MpmcPublish() : _readCursor(0)
+    MpmcPublish() : _readPointer(nullptr)
     {
-        static_assert(MAX_THREADS > 0 && MAX_THREADS < TOO_MANY_THREADS,
-            "entries in MpmcPublish cannot exceed TOO_MANY_THREADS");
+        static_assert(MAX_THREADS > 2 && MAX_THREADS < 512, "entries must > 2 < 512 in MpmcPublish");
     }
 
     ~MpmcPublish() {}
 
     // return false if empty
     bool read(T& target) const
     {
-        if (empty())
-        {
-            return false;
-        }
-
-        for (;;)
+        for (auto* cell = _readPointer.load(std::memory_order_consume); cell;
```

Comment:
Simplified a bit. Most likely outcome is that we get a readable item with ref count > 0.
One less load of the _readPointer as we check for empty from same read.